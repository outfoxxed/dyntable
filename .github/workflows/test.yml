name: Run Tests
on: push
jobs:
  test:
    name: Run Tests
    runs-on: ubuntu-latest
    steps:
      - name: checkout
        uses: actions/checkout@v3
      - name: setup nix
        uses: cachix/install-nix-action@v19

      - name: check formatting
        run: nix develop .# -c cargo fmt --check
      - name: check default features
        run: nix develop .#stable -c cargo check
      - name: check allocator_api
        run: nix develop .# -c cargo check --features allocator_api
      - name: run stable tests
        run: nix develop .#stable -c cargo test
      - name: run tests under miri
        run: nix develop .# -c cargo miri test
