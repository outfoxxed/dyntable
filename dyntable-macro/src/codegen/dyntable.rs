use proc_macro2::TokenStream;
use quote::format_ident;
use syn::{Signature, Token};

use crate::parse::{DynTraitInfo, VTableEntry};

pub fn impl_dyntable(dyntrait: &DynTraitInfo) -> TokenStream {
	let proxy_trait = format_ident!("__DynTable_{}", dyntrait.dyntrait.ident);
	let ident = &dyntrait.dyntrait.ident;
	let vtable_ident = &dyntrait.vtable.name;
	let (_, ty_generics, where_clause) = dyntrait.generics.split_for_impl();
	let impl_generics = dyntrait.generics.params.clone().into_iter()
    .collect::<Vec<_>>();
	let where_entries = where_clause
    .into_iter()
    .flat_map(|clause| &clause.predicates);

	let vtable_entries = dyntrait.entries.iter()
    .map(|entry| match entry {
			VTableEntry::Subtable(subtable) => {
				let entry_name = &subtable.ident;
				quote::quote! {
					#entry_name: <__DynTarget as ::dyntable::DynTable<
						<(dyn #ident #ty_generics + 'static) as ::dyntable::VTableRepr>::VTable,
					>>::VTABLE
				}
			},
			VTableEntry::Method(method) => {
				let Signature {
					unsafety,
					abi,
					fn_token,
					ident,
					inputs,
					variadic,
					output,
					..
				} = &method.sig;

				let inputs = inputs.iter()
					.map(|_| <Token![_]>::default());

				let output = match output {
					syn::ReturnType::Default => None,
					syn::ReturnType::Type(..) => Some(<Token![_]>::default()),
				}.into_iter();

				quote::quote! {
					#ident: unsafe {
						::core::intrinsics::transmute(
							__DynTarget::#ident as
								#unsafety #abi #fn_token (
									#(#inputs,)* #variadic
								) #( -> #output)*
						)
					}
				}
			},
		});

	quote::quote! {
		#[allow(non_camel_case_types)]
		unsafe trait #proxy_trait<'v, V: 'v + ::dyntable::VTable> {
			const VTABLE: V;
			const STATIC_VTABLE: &'v V;
		}

		unsafe impl<
			'__dyn_vtable,
			#(#impl_generics + '__dyn_vtable,)*
			__DynTarget,
		> ::dyntable::__private::DynTable2<'__dyn_vtable, #vtable_ident #ty_generics>
		for ::dyntable::__private::DynImplTarget<__DynTarget, #vtable_ident #ty_generics>
		where
			#(#where_entries,)*
			__DynTarget: #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>,
		{
			const VTABLE: #vtable_ident #ty_generics = __DynTarget::VTABLE;
			const STATIC_VTABLE: &'__dyn_vtable #vtable_ident #ty_generics = __DynTarget::STATIC_VTABLE;
		}

		unsafe impl<
			'__dyn_vtable,
			#(#impl_generics + '__dyn_vtable,)*
			__DynTarget: #ident #ty_generics,
		> #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>
		for __DynTarget
		#where_clause {
			const STATIC_VTABLE: &'__dyn_vtable #vtable_ident #ty_generics =
				&<Self as #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>>::VTABLE;
			const VTABLE: #vtable_ident #ty_generics = #vtable_ident {
				#(#vtable_entries,)*
			};
		}
	}
}
