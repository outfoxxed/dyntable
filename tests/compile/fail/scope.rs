use test::{PrivTrait, PrivTraitVTable};

fn main() {}

mod test {
	use dyntable::dyntable;

	#[dyntable]
	trait PrivTrait {
		extern "C" fn test(&self);
	}
}
