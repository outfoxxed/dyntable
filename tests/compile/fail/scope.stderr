error[E0603]: trait `PrivTrait` is private
 --> tests/compile/fail/scope.rs:1:12
  |
1 | use test::{PrivTrait, PrivTraitVTable};
  |            ^^^^^^^^^ private trait
  |
note: the trait `PrivTrait` is defined here
 --> tests/compile/fail/scope.rs:8:2
  |
8 |     #[dyntable]
  |     ^^^^^^^^^^^
  = note: this error originates in the attribute macro `dyntable` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0603]: struct `PrivTraitVTable` is private
 --> tests/compile/fail/scope.rs:1:23
  |
1 | use test::{PrivTrait, PrivTraitVTable};
  |                       ^^^^^^^^^^^^^^^ private struct
  |
note: the struct `PrivTraitVTable` is defined here
 --> tests/compile/fail/scope.rs:8:2
  |
8 |     #[dyntable]
  |     ^^^^^^^^^^^
  = note: this error originates in the attribute macro `dyntable` (in Nightly builds, run with -Z macro-backtrace for more info)
